source/lac/scalapack.cc

```diff
@@ -0,0 +1,966 @@
+
+    // Below we always try to create 2D processor grids:
+
+    int n_processes;
+    MPI_Comm_size(mpi_comm, &n_processes);
+
+    // Get the total number of cores we can occupy in a rectangular dense matrix
+    // with rectangular blocks when every core owns only a single block:
+    const int n_processes_heuristic = int(std::ceil((1.*m)/block_size_m))*
+                                      int(std::ceil((1.*n)/block_size_n));
+    const int Np = std::min(n_processes_heuristic, n_processes);
+
+    // Now we need to split Np into  Pr x Pc. Assume we know the shape/ratio
+    // Pc =: ratio * Pr
+    // therefore
+    // Np = Pc * Pc / ratio
+    // for quadratic matrices the ratio equals 1
+    const double ratio = double(n)/m;
+    int Pc = std::floor(std::sqrt(ratio * Np));
+
+    // one could rounds up Pc to the number which has zero remainder from the division of Np
+    // while ( Np % Pc != 0 )
+    //  ++Pc;
+    // but this affects the grid shape dramatically, i.e. 10 cores 3x3 becomes 2x5.
+
+    // limit our estimate to be in [2, Np]
+    int n_process_columns = std::min (Np, std::max(2, Pc));
+    // finally, get the rows:
+    int n_process_rows = Np / n_process_columns ;
+
+    Assert (n_process_columns >=1 && n_process_rows >=1 && n_processes >= n_process_rows*n_process_columns,
+            ExcMessage("error in process grid: "+
+                       std::to_string(n_process_rows)+"x"+
+                       std::to_string(n_process_columns)+
+                       "="+
+                       std::to_string(n_process_rows*n_process_columns)+
+                       " out of " +
+                       std::to_string(n_processes)));
+
+    return std::make_pair(n_process_rows,n_process_columns);
+
+    // For example,
+    // 320x320 with 32x32 blocks and 16 cores:
+    // Pc = 1.0 * Pr  => 4x4 grid
+    // Pc = 0.5 * Pr  => 8x2 grid
+    // Pc = 2.0 * Pr  => 3x5 grid
+  }
+}
+
+
+
+ProcessGrid::ProcessGrid(MPI_Comm mpi_comm,
+                         const std::pair<unsigned int,unsigned int> &grid_dimensions)
+  :
+  mpi_communicator(mpi_comm),
+  this_mpi_process(Utilities::MPI::this_mpi_process(mpi_communicator)),
+  n_mpi_processes(Utilities::MPI::n_mpi_processes(mpi_communicator)),
+  n_process_rows(grid_dimensions.first),
+  n_process_columns(grid_dimensions.second)
+{
+  Assert (grid_dimensions.first > 0,
+          ExcMessage("Number of process grid rows has to be positive."));
+  Assert (grid_dimensions.second > 0,
+          ExcMessage("Number of process grid columns has to be positive."));
+
+  Assert (grid_dimensions.first*grid_dimensions.second <= n_mpi_processes,
+          ExcMessage("Size of process grid is larger than number of available MPI processes."));
+
+  // processor grid order.
+  const bool column_major = false;
+
+  // Initialize Cblas context from the provided communicator
+  blacs_context = Csys2blacs_handle(mpi_communicator);
+  const char *order = ( column_major ? "Col" : "Row" );
```

Comment:
Why do you do this? `column_major` is always `false` so it will always be `Row`.

Reply:
just to have a quick switch if needed.